src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5291,12 +5295,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           // however we cannot do `context.withMacrosDisabled`
           // because `expr` might contain nested macro calls (see scala/bug#6673)
           //
-          // Note: apparently `Function(Nil, EmptyTree)` is the secret parser marker
+          // Note: `Function(Nil, EmptyTree)` is the secret parser marker
           // which means trailing underscore -- denoting a method value. See makeMethodValue in TreeBuilder.
-          case Typed(expr, Function(Nil, EmptyTree)) =>
+          case MethodValue(expr) =>
             typed1(suppressMacroExpansion(expr), mode, pt) match {
               case macroDef if treeInfo.isMacroApplication(macroDef) => MacroEtaError(macroDef)
-              case methodValue                                       => typedEta(checkDead(methodValue), expr)
+              case methodValue                                       => typedEta(checkDead(methodValue), tree)
```

Comment:
So this is crux of the change. Can you describe how the surrounding changes allowed you go back to passing the entire `Typed` tree as `original` here, without incurring the infinite loop I was firefighting in c15b9b730e76f2c33399ce4ffc2ffb3c71025531?

Reply:
I probably can after I look at the code again.